src/sentry/ratelimits/cardinality.py

```diff
@@ -132,3 +157,266 @@ def use_quotas(
+       mismatches between code and DB that somehow don't surface during the
+       DB read). In that case the quotas would be consumed immediately and
+       incident recovery would require us to reset all quotas manually (or
+       disable rate limiting via some killswitch)
+
+    3) The redis backend (really the only backend we care about) already
+       has some consistency problems.
+
+       a) Redis only provides strong consistency and ability to
+          check-and-increment counters when all involved keys hit the same
+          Redis node. That means that a quota with prefix="org_id:123" can
+          only run on a single redis node. It also means that a global
+          quota (`prefix="global"`) would have to run on a single
+          (unchangeable) redis node to be strongly consistent. That's
+          however a problem for scalability.
+
+          There's no obvious way to make global quotas consistent with
+          per-org quotas this way, so right now it means that requests
+          containing multiple quotas with different `prefixes` cannot be
+          checked-and-incremented atomically even if we were to change the
+          rate-limiter's interface.
+
+       b) This is easily fixable, but because of the above, we
+          currently don't control Redis sharding at all, meaning that even
+          keys within a single quota's window will hit different Redis
+          node. This also helps further distribute the load internally.
+
+          Since we have given up on atomic check-and-increments in general
+          anyway, there's no reason to explicitly control sharding.
+    """
+
+    def __init__(
+        self,
+        cluster: str = "default",
+        cluster_num_shards: int = 3,
+        cluster_num_physical_shards: int = 3,
+    ) -> None:
+        """
+        :param cluster: Name of the redis cluster to use, to be configured with
+            the `redis.clusters` Sentry option (like any other redis cluster in
+            Sentry).
+        :param cluster_num_shards: The number of logical shards to have. This
+            controls the average set size in Redis.
+        :param cluster_num_physical_shards: The number of actual shards to
+            store. Controls how many keys of type "unordered set" there are in
+            Redis. The ratio `cluster_num_physical_shards / cluster_num_shards`
+            is a sampling rate, the lower it is, the less precise accounting
+            will be.
+        """
+        self.client = redis.redis_clusters.get(cluster)
+        assert 0 < cluster_num_physical_shards <= cluster_num_shards
+        self.cluster_num_shards = cluster_num_shards
+        self.cluster_num_physical_shards = cluster_num_physical_shards
+        super().__init__()
+
+    @staticmethod
+    def _get_timeseries_key(request: RequestedQuota, hash: Hash) -> str:
+        return f"cardinality:counter:{request.prefix}-{hash}"
+
+    def _get_read_sets_keys(
```

Comment:
These methods also become simpler if you hardcode one physical shard.

Reply:
see the above response, i think this would become annoying if we ever scale beyond our typical ring of three redis nodes